core/src/main/java/com/datastax/oss/driver/api/core/metadata/diagnostic/NodeGroupDiagnostic.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.metadata.diagnostic;
+
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Map;
+
+/**
+ * A health {@link Diagnostic} for a group of nodes, detailing how many nodes were found in total,
+ * how many were up, and how many were down.
+ *
+ * <p>The breadth of the report's scope depends on how it was created; it may refer to the entire
+ * cluster, or to just a datacenter.
+ */
+public interface NodeGroupDiagnostic extends Diagnostic {
+
+  /** @return the total number of nodes in this group. */
+  int getTotal();
+
+  /**
+   * Returns the number of nodes in this group whose {@linkplain
+   * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+   * com.datastax.oss.driver.api.core.metadata.NodeState#UP UP}.
+   *
+   * @return the number of nodes in this group that are known to be up.
+   */
+  int getUp();
+
+  /**
+   * Returns the number of nodes in this group whose {@linkplain
+   * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+   * com.datastax.oss.driver.api.core.metadata.NodeState#DOWN DOWN} or {@link
+   * com.datastax.oss.driver.api.core.metadata.NodeState#FORCED_DOWN FORCED_DOWN}.
+   *
+   * @return the number of nodes in this group that are known to be down.
+   */
+  int getDown();
+  /**
+   * Returns the number of nodes in this group whose {@linkplain
+   * com.datastax.oss.driver.api.core.metadata.Node#getState() state} is {@link
+   * com.datastax.oss.driver.api.core.metadata.NodeState#UNKNOWN UNKNOWN}.
+   *
+   * <p>Nodes may be in an unknown state if the driver hasn't connected to them at all, and it has
+   * not received any Gossip event indicating their actual state. Most of such nodes should actually
+   * be up.
+   *
+   * @return the number of nodes in this group that are in unknown state.
+   */
+  int getUnknown();
+
+  /**
+   * Returns the status of this node group. The status will be {@link Status#AVAILABLE} if all nodes
+   * are up; {@link Status#UNAVAILABLE} if all nodes are down, or if the group has no node at all.
+   * In all other cases the status will be {@link Status#PARTIALLY_AVAILABLE}.
+   */
+  @NonNull
+  @Override
+  default Status getStatus() {
+    if (getTotal() == 0 || getDown() == getTotal()) {
```

Comment:
Are we providing happens-before relation here? I mean that the `getTotal() == 0` may not necessary mean down. It may be the situation when the `getStatus()` was executed before there was underlying health check that incremented `getTotal`. Or is it not possible? Maybe when `getTotal()==0` we should return UNKNOWN state?

Reply:
> that the getTotal() == 0 may not necessary mean down

yes it does, cf. javadocs: 

```
The status will be {@link Status#AVAILABLE} if all nodes are up; {@link Status#UNAVAILABLE} if all nodes are down, or if the group has no node at all.
```

> It may be the situation when the getStatus() was executed before there was underlying health check that incremented getTotal.

That shouldn't be allowed, and our implementations are all immutable, so this cannot happen.